README.md
# C++ Stub repository

A stub repository for C++ development projects in W-13.

This repository will contain the necessary setup files to integrate C++ doc strings,
[CMake](https://cmake.org/cmake/help/v3.14/), [Doxygen](https://www.doxygen.nl/manual/docblocks.html),
[Sphinx](https://www.sphinx-doc.org/en/master/), and [Breathe](https://breathe.readthedocs.io/en/latest/) for a complete
build system with integrated documentation. It will also include the necessary hooks to commonly used C++ libraries for
constitutive modeling. This stub repository also includes template hooks for integrating C++ code as Abaqus subroutines.

> Note: you can use this repo as a stub for fortran projects as well! A
> step-by-step may never happen because c++ is the future of W-13 subroutines.
> For now, refer to the following references:
>
> * [CMake for Fortran example
>   ``CMakeLists.txt``](https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/languages/fortran/ForFortranExample)
> * [CMake documentation starting point](https://cmake.org/cmake/help/v3.14/module/CheckFortranSourceRuns.html)
> * [Stack Overflow thread](https://stackoverflow.com/questions/12705562/using-cmake-with-fortran)
> * [Doxygen comments for Fortran](https://www.doxygen.nl/manual/docblocks.html#fortranblocks)

## Setting up a new project from this stub repo

This section has a corresponding section in the formatted and hyperlinked html
documentation. The html documentation also includes setup instructions for
Jenkins Continuous Integration (CI) setup. Build and view instructions are
included in following sections, separated from project setup by horizontal bars.

### Prerequisites

Set up your profile on Bitbucket with ssh keys. You can follow the instructions
on the W-13 Confluence page [Gitting started with W-13
Bitbucket](https://xcp-confluence.lanl.gov/display/GIT/Gitting+Started+W-13%27s+Git+Server)

### Clone cpp\_stub into a local repository

1. Navigate to the [upstream cpp\_stub repo](https://xcp-stash.lanl.gov/projects/MM/repos/cpp_stub/browse)

2. Copy the ssh url from the Bitbucket "Clone" button on the Bitbucket
   repository web page. It should look like the following:

       ssh://git@xcp-stash.lanl.gov:7999/mm/cpp_stub.git

3. Navigate to your preferred repository directory in a terminal or use the
   example commands below

       $ ssh -X sstelmo.lanl.gov
       $ pwd
       /home/<moniker>
       $ mkdir -p /projects/$USER/w13repos
       $ cd /projects/$USER/w13repos
       $ pwd
       /projects/<moniker>/w13repos

4. Clone the stub repository using the url copied in step 2.

       $ pwd
       /projects/<moniker>/w13repos
       $ git clone ssh://git@xcp-stash.lanl.gov:7999/mm/cpp_stub.git

5. Rename the local repository directory for your project.

       $ pwd
       /projects/<moniker>/w13repos
       $ ls cpp_stub -d
       cpp_stub
       $ mv cpp_stub my_project
       $ ls cpp_stub -d
       ls: cannot access 'cpp_stub': No such file or directory
       $ mv cpp_stub my_project
       my_project

6. Change to your project's repository directory

       $ pwd
       /projects/<moniker>/w13repos
       $ cd my_project
       $ pwd
       /projects/<moniker>/w13repos/my_project

### Create a new repository on Bitbucket

> Note: These notes are a text copy of a variation on the [New Bitbucket Repo
> Guide](https://simulia.lanl.gov/ECMF-D/devops_guide.html#new-bitbucket-repo-guide)
> which can also be found in the [W-13 DevOps
> Manual](https://xcp-confluence.lanl.gov/display/COM/W-13+DevOps)

1. Navigate to the W-13 [Material Models](https://xcp-stash.lanl.gov/projects/MM) Bitbucket project.

2. Create a new repository by clicking on the "+" sign in the upper left corner.

3. Enter a name for your project and click "Create repository"

4. Follow the "My code is already tracked by Git" instructions.

       $ pwd
       /projects/<moniker>/w13repos/my_project
       $ git remote set-url origin ssh://git@xcp-stash.lanl.gov:7999/mm/my_project.git
       $ git push -u origin --all
       $ git push origin --tags

5. Refresh the Bitbucket webpage and verify that the repository code was pushed
   correctly. You should see a list of source files and this Bitbucket parsed
   ``README.md`` displayed. You can also select the drop down branch menu to
   view a "master" and "dev" branch.

### Update settings for your repository

Bitbucket repositories in the [Material
Models](https://xcp-stash.lanl.gov/projects/MM) project inherit permissions and
settings from that project. This included read permission for the
``w13bitbucket`` [UNIX
group](https://xcp-confluence.lanl.gov/pages/viewpage.action?pageId=150929410&searchId=Y1EVB37UN).
For most developers, these inherited repository settings are appropriate and
only a small number of settings must be updated.

1. Click on the gear icon in the lower left sidebar.

2. From the "Repository details" landing page, update the default branch from
   "master" to "dev".

3. From the "Repository permissions" tab you can add additional permissions by
   user and UNIX group.

4. From the "Default reviewers" tab you can add yourself and any project
   co-owners as default reviewers.

### Fork the upstream repository

In the [Forking
Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow)
the repository you just created in the [Material
Models](https://xcp-stash.lanl.gov/projects/MM) project is called the "upstream"
repository. Throughout older W-13 documentation this may also be called the
"official" repository.

Bitbucket repositories that inherit permissions from W-13 projects use the
[Forking
Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow)
and limit permissions for pushing changes to the upstream repository. Now that
branches exist on this repository, no one will be able to push directly to
*existing branches* of the upstream respository.

1. Click the fork button in the left hand sidebar just above the gear icon.

2. Click "fork this repository" button.

3. Verify

   a. The "Project" points to your personal project space in Bitbucket. It will
      probably show your full name as it appears in the phonebook.

   b. The "Name" matches your project name

   c. The "Enable fork syncing" checkbox is checked

4. Click the "Fork Repository" button. You should land on a familiar looking
   source code view of your repository, but now located in your personal project
   space.

This repository is referred to as the "fork" or "remote" repository throughout
W-13 DevOps documentation.

### Update the remote url in your local repository

The final repo setup step is to update the remote url of the local clone of
``my_project``.  We will return to the terminal session.

1. Copy the url of your "remote" repository from the Bitbucket webpage. It
   should look like:

       ssh://git@xcp-stash.lanl.gov:7999/~<moniker>/my_project.git

2. Return to your terminal session and update the remote repository for the
   final time.

       $ pwd
       /projects/<moniker>/w13repos/my_project
       $ git remote set-url origin ssh://git@xcp-stash.lanl.gov:7999/~<moniker>/my_project.git
       $ git push -u origin --all
       $ git push origin --tags

### Update project name throughout repository

> Note: the remaining steps are a truncated version of the W-13 Git project
> [contribution guide](https://simulia.lanl.gov/ECMF-D/contribution_guide.html)
> which can also be found in the [W-13 DevOps
> Manual](https://xcp-confluence.lanl.gov/display/COM/W-13+DevOps). Critically,
> these steps will omit the Jira task creation and Bitbucket Pull-Request (PR)
> steps. The Bitbucket PR steps may be reproduced using the contribution guide,
> but your project will have to create a Jira project prior to integrating the
> Jira workflow. Contact the xcp devops team <devops-help@lanl.gov> to create a
> Jira project. You can email the W-13 DevOps team <w13devops@lanl.gov> for
> notes about setup.

1. Create a feature branch for your project name updates

       $ pwd
       /projects/<moniker>/w13repos/my_project
       $ git checkout -b feature/project-name-updates
       $ git branch
         dev
       * feature/project-name-updates
         master

2. Search for all instances of ``cpp_stub``. The list of occurrences will look
   quite long, but we can search and replace with ``sed`` to avoid manual file
   edits.

       $ pwd
       /projects/<moniker>/w13repos/my_project

       # Recursive, case-insensitive search and count occurrences
       $ grep -ri cpp_stub . --exclude-dir={build,.git} | wc -l
       57

       # Recursive, case-insensitive search and display
       $ grep -ri cpp_stub . --exclude-dir={build,.git}
       ...

       # Clean list of files with project name
       $ grep -ri cpp_stub . --exclude-dir={build,.git} -l
       ./CMakeLists.txt
       ./docs/api.rst
       ./docs/devops.rst
       ./README.md
       ./set_vars.sh
       ./src/cpp/cpp_stub.cpp
       ./src/cpp/cpp_stub.h
       ./src/cpp/tests/test_cpp_stub.cpp

3. Search and replace from command line

       $ pwd
       /projects/<moniker>/w13repos/my_project

       # Replace lower case occurrences in place
       $ sed -i 's/cpp_stub/my_project/g' $(grep -ri cpp_stub . --exclude-dir={build,.git} -l)
       $ grep -ri cpp_stub . --exclude-dir={build,.git} -l
       ./src/cpp/cpp_stub.h

       # Replace upper case occurrences in place
       $ sed -i 's/CPP_STUB/MY_PROJECT/g' $(grep -ri cpp_stub . --exclude-dir={build,.git} -l)

4. Verify no more occurrences of project name ``cpp_stub``

       $ pwd
       /projects/<moniker>/w13repos/my_project
       $ grep -ri cpp_stub . --exclude-dir={build,.git} | wc -l
       0
       $ grep -ri cpp_stub . --exclude-dir={build,.git}
       # no stdout to terminal because no files found
       $ grep -ri cpp_stub . --exclude-dir={build,.git} -l
       # no stdout to terminal because no files found

5. Search and replace camelcase project name occurrences, e.g. ``cppStub``.

       $ grep -r cppStub . --exclude-dir={build,.git}
       ...
       $ sed -i 's/cppStub/myProject/g' $(grep -r cppStub . --exclude-dir={build,.git} -l)
       $ grep -r cppStub . --exclude-dir={build,.git} -l
       # no stdout to terminal because no files found

6. Find files containing the project in their file name

       $ pwd
       /projects/<moniker>/w13repos/my_project
       $ find . -type d \( -name .git -o -name build \) -prune -false -o -name "*cpp_stub*"
       ./src/cpp/cpp_stub.cpp
       ./src/cpp/cpp_stub.h
       ./src/cpp/tests/test_cpp_stub.cpp

7. Rename files after current project

       $ rename 's/cpp_stub/myproject/' $(find . -type d \( -name .git -o -name build \) -prune -false -o -name "*cpp_stub*")

8. Commit and push your changes to your "remote" or "fork" repository

       $ pwd
       /projects/<moniker>/w13repos/my_project
       # Add tracked files and message
       $ git commit -a -m "FEAT: replace cpp_stub with my_project through repository"
       $ git push origin feature/project-name-updates

You can also perform some cleanup in ``README.md`` to remove this walk-through.

From here, the W-13 best practice workflow would return to the Bitbucket webpage
and submit a Pull-Request from the ``feature/project-name-updates`` branch of
``\<moniker\>/my_project`` repository (a.k.a. fork or remote) to the ``dev`` branch
of ``Material Models/my_project`` repository (a.k.a. upstream or official).

After updating your project by merging to the upstream repository, the fork
syncing feature of Bitbucket will automatically update any identically named
branches in your fork repository. Best practices suggest you should limit the
upstream repository branches to clean ``dev`` and ``master`` branches and
*NEVER* develop directly on the ``dev`` and ``master`` branches of your fork
repository. Limit development work to ``feature/thing`` type branches on your
fork/remote repo and frequently commit changes and push from the local feature
branch back to the fork/remote repo.

Happy hacking!

## Setting up a Jenkins CI job

This section has a corresponding section in the formatted and hyperlinked html
documentation. Build and view instructions are included in following sections,
separated from project setup by horizontal bars.

See ``docs/jenkins_job_creation.md`` in this repo.

A quick build of just the jenkins markdown file can be performed with ``pandoc``

    $ pwd
    /path/to/local/cpp_stub
    $ pandoc docs/jenkins_job_creation.md -o jenkins_job_creation.html

## Version control for a Jenkins job

This section has a corresponding section in the formatted and hyperlinked html
documentation. Build and view instructions are included in following sections,
separated from project setup by horizontal bars.

See ``docs/jenkins_job_xml.md`` in this repo.

A quick build of just the jenkins markdown file can be performed with ``pandoc``

    $ pwd
    /path/to/local/cpp_stub
    $ pandoc docs/jenkins_job_xml.md -o jenkins_job_xml.html

---

## Dependencies

### Compilers

* c++11 compiler (listed version number has been tested at some point)

  * g++ >= GNU 4.8.5

### Executables

* [CMake](https://cmake.org/cmake/help/v3.14/) >= 3.14
* [Doxygen](https://www.doxygen.nl/manual/docblocks.html) >= 1.8.5
* [LaTeX](https://www.latex-project.org/help/documentation/) >= 2017

### Python Modules (for documentation)

For convenience, the minimal Python environment requirements for the
documentation build are included in ``configuration_files/environment.yaml``.
This file was created from the [pipreqs](https://github.com/bndr/pipreqs)
command line tool and Sphinx configuration inspection, e.g. the extension
packages.

    $ pwd
    path/to/cpp_stub/
    $ pipreqs --use-local --print --no-pin .

A minimal anaconda environment for building the documentation can be created
from an existing anaconda installation with the following commands.

    $ conda env create --file configuration_files/environment.yaml

You can learn more about Anaconda Python environment creation and management in
the [Anaconda
Documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html)

### C++ Libraries

> **NOTE: Non-admin installations for Eigen and Boost are no longer required.** This project is built and deployed
> against C++ libraries managed in Conda. See the Conda environment file and README discussion for non-admin environment
> management.

* [Eigen](https://eigen.tuxfamily.org/dox/) >= 3.3.7
* [BOOST](https://www.boost.org/doc/libs/1_53_0/) >= 1.59.0
* error\_tools: https://xcp-stash.lanl.gov/projects/MM/repos/error_tools
* vector\_tools: https://xcp-stash.lanl.gov/projects/MM/repos/vector_tools
* stress\_tools: https://xcp-stash.lanl.gov/projects/MM/repos/stress_tools
* solver\_tools: https://xcp-stash.lanl.gov/projects/MM/repos/solver_tools
* constitutive\_tools: https://xcp-stash.lanl.gov/projects/MM/repos/consitutive_tools

If not found on the current system or active Conda environment, all of the
``*_tools`` libraries are pulled from their git repos by branch name and built
with their respective cmake files as part of the cmake build for this project.

---

## Build and Test

This project is built with [CMake](https://cmake.org/cmake/help/v3.14/) and uses
[Sphinx](https://www.sphinx-doc.org/en/master/) to build the documentation with
[Doxygen](https://www.doxygen.nl/manual/docblocks.html) +
[Breathe](https://breathe.readthedocs.io/en/latest/) for the c++ API.

### Build on sstelmo

1) Activate the correct python environment

       $ module load python/2020.07-python-3.8
       $ sv3r

2) Create a build directory

       $ pwd
       /path/to/cpp_stub/

       $ mkdir build
       $ cd build

3) Configure ``cmake3``

   > This step only needs to be performed once unless you need to specify a new CMake configuration for a re-build. Most
   > command line arguments and environment variables are stored in the CMake cache. Anything found in cache will not be
   > re-configured unless you remove the cache file or clobber the build directory.

       $ pwd
       /path/to/cpp_stub/build
       $ cmake3 ..

4) Build various portions of the project

   > Most of the project will re-build only as necessary after source updates. Some portions of the documentation
   > require a ``make clean`` after documentation source file updates to force a re-build.

       $ pwd
       /path/to/cpp_stub/build

       # Build everything
       $ cmake3 --build .

       # Build only the c++ primary libraries
       $ cmake3 --build src/cpp

5) Locate build files

   > The build directory structure may change between version releases. Developers and users are encouraged to become
   > familiar with the bash ``find``, ``grep``, and ``tree`` commands to locate build files.

       $ pwd
       /path/to/cpp_stub/build

       # find c++ libraries and ignore intermediate files with similar extensions
       $ find . \( -name "*.o" -o -name "*.so" -o -name "*.a" \) | grep -vE "\.cpp\."

6) Clean build directory to force a re-build

   > :warning: :warning: :warning: HEALTH WARNING :warning: :warning: :warning:
   >
   > The abaqus input files and bash scripts used for integration testing are built with the [CMake
   > ``add_custom_target``](https://cmake.org/cmake/help/latest/command/add_custom_target.html) feature. Consequently, the
   > integration test target is _always considered out of date_. The integration test target copies all registered input
   > files and the integration test bash script from source to build directory. This means the file copy operation is
   > always performed when the integration test target is requested in the cmake build command, e.g. ``cmake --build .``
   > or ``cmake --build src/abaqus/tests``. This operation is computationally inexpensive with respect to building the
   > ``cpp_stub`` source code.
   >
   > Input files are registered in the ``src/abaqus/tests/CMakeLists.txt`` file under the ``ABAQUS_INPUT_FILES`` CMake
   > variable.


       $ pwd
       /path/to/cpp_stub/build

       $ make clean

### Test on sstelmo

4) Build tests of the project

       $ pwd
       /path/to/cpp_stub/build

       # Build c++ tests
       $ cmake3 --build src/cpp/tests

       # Build Abaqus integration tests
       $ cmake3 --build src/abaqus/tests

5) Run the tests

       $ pwd
       /path/to/cpp_stub/build

       # Run ctest
       $ ctest

       # Results print to screen
       # View details of most recent test execution including failure messages
       $ less Testing/Temporary/LastTest.log

### Convenience build wrappers

Two build scripts have been created for convenience, ``new_build.sh`` and
``build_docs.sh``. The first will build everything including the library binary,
the test binary, and the documentation. This is the same build script used by
``jenkins_build.sh`` for CI builds and testing. The ``build_docs.sh`` script
only builds the documentation. Both build scripts clobber existing build
directories, reset any bash environment variables, and run the cmake
configuration from scratch.

2) Build everything and run tests

       $ pwd
       /path/to/cpp_stub/

       # Just perform the build (pick one)
       $ ./new_build.sh <cmake build type>
       $ ./new_build.sh None
       $ ./new_build.sh Release

       # Perform tests from PWD
       $ ./build/src/cpp/tests/test_cpp_stub

       # Build and perform tests
       $ ./jenkins_build.sh

3) View test results

       # As built directly to PWD
       $ cat results.tex

       # As built by jenkins_build.sh
       $ cat build/src/cpp/tests/*_results.tex
       $ cat *results.tex

4) Display docs

       # Sphinx
       $ firefox build/docs/sphinx/html/index.html &

       # Doxygen
       $ firefox build/docs/doxygen/html/index.html &

### Building the documentation

> :warning: :warning: :warning: HEALTH WARNING :warning: :warning: :warning:
>
> **API Health Note**: The sphinx API docs are a work-in-progress. The doxygen
> API is much more useful.

The documentation can be built with ``build_docs.sh``. The steps used in that
shell script are repeated here.

To build just the documentation pick up the steps here:

2) Create the build directory and move there

       $ pwd
       /path/to/cpp_stub/
       $ mkdir build/
       $ cd build/

3) Run cmake3 configuration

       $ pwd
       /path/to/cpp_stub/build/
       $ cmake3 ..

4) Build the docs

       $ cmake3 --build docs

5) Documentation builds to:

       cpp_stub/build/docs/sphinx/html/index.html

6) Display docs

       $ pwd
       /path/to/cpp_stub/build/
       $ firefox docs/sphinx/html/index.html &

7) While the Sphinx API is still a WIP, try the doxygen API

       $ pwd
       /path/to/cpp_stub/build/
       $ firefox docs/doxygen/html/index.html &

## Install the library

Build the entire before performing the installation.

4) Build the entire project

       $ pwd
       /path/to/cpp_stub/build
       $ cmake3 --build .

5) Install the library

       $ pwd
       /path/to/cpp_stub/build
       $ cmake --install . --prefix path/to/root/install

       # Example local user (non-admin) Linux install
       $ cmake --install . --prefix /home/$USER/.local

       # Example install to conda environment
       $ conda active my_env
       $ cmake --install . --prefix ${CONDA_DEFAULT_ENV} 

       # Example install to W-13 CI/CD conda environment performed by CI/CD institutional account
       $ cmake --install . --prefix /projects/python/release

---

## Contribution Guidelines

### Git Commit Message

Begin Git commit messages with one of the following headings:

* BUG: bug fix
* DOC: documentation
* FEAT: feature
* MAINT: maintenance
* TST: tests
* REL: release
* WIP: work-in-progress

For example:

    git commit -m "FEAT: short intent of new feature"
    git commit -m "BUG: fixes nasty bug"
    git commit -m "DOC: adds documentation for feature"

### Git Branch Names

When creating branches use one of the following naming conventions. When in
doubt use ``feature/<description>``.

* bugfix/\<description>
* feature/\<description>
* release/\<description>

### reStructured Text

[Sphinx](https://www.sphinx-doc.org/en/master/) reads in docstrings and other special portions of the code as
reStructured text. Developers should follow styles in this [Sphinx style
guide](https://documentation-style-guide-sphinx.readthedocs.io/en/latest/style-guide.html#).

### Style Guide

This project does not yet have a full style guide. Generally, wherever a style can't be
inferred from surrounding code this project falls back to
[PEP-8](https://www.python.org/dev/peps/pep-0008/)-like styles. There are two
notable exceptions to the notional PEP-8 fall back:

1. [Doxygen](https://www.doxygen.nl/manual/docblocks.html) style docstrings are
   required for automated, API from source documentation.
2. This project prefers expansive whitespace surrounding parentheses, braces, and
   brackets.

   * No leading space between a function and the argument list.
   * One space following an open paranthesis ``(``, brace ``{``, or bracket
     ``[``
   * One space leading a close paranthesis ``)``, brace ``}``, or bracket ``]``

An example of the whitespace style:

    my_function( arg1, { arg2, arg3 }, arg4 );

The following ``sed`` commands may be useful for updating white space, but must
be used with care. The developer is recommended to use a unique git commit
between each command with a corresponding review of the changes and a unit test
run.

* Trailing space for open paren/brace/bracket

      sed -i 's/\([({[]\)\([^ ]\)/\1 \2/g' <list of files to update>

* Leading space for close paren/brace/bracket

      sed -i 's/\([^ ]\)\([)}\]]\)/\1 \2/g' <list of files to update>

* White space between adjacent paren/brace/bracket

      sed -i 's/\([)}\]]\)\([)}\]]\)/\1 \2/g' <list of files to update>


docs/changelog.rst
.. _changelog:


#########
Changelog
#########

******************
0.1.5 (unreleased)
******************

Documentation
=============
- Update project setup instructions from Atlassian to Gitlab workflows (:issue:`2`, :merge:`4`). By `Kyle Brindley`_.

Internal Changes
================
- Convert README from markdown to restructured text (:issue:`2`, :merge:`4`). By `Kyle Brindley`_.
- Separate Abaqus integration test setup from Abaqus integration ctest declaration. Enables documentation build
  dependencies on Abaqus integration test input files without requiring Abaqus test execution on systems with no Abaqus
  installation (:issue:`2`, :merge:`4`). By `Kyle Brindley`_.


******************
0.1.4 (2021-07-13)
******************

Internal Changes
================
- Upstream project settings update to set default merge-request branch. By `Kyle Brindley`_.

******************
0.1.3 (2021-07-13)
******************

- Migrate from ddw-bibucket.lanl.gov to re-git.lanl.gov and convert to Gitlab CI/CD (:issue:`1`, :merge:`1`). By `Kyle
  Brindley`_.

******************
0.1.2 (2021-07-01)
******************

Internal Changes
================
- Use Git SCM tags for semantic versioning (:jira:`702`, :pull:`50`). By `Kyle Brindley`_.
- Master branch production release logic for CD, including automated micro-version bumps (:jira:`702`, :pull:`50`). By `Kyle
  Brindley`_.


******************
0.1.1 (2021-06-15)
******************

Bug Fixes
=========
- Corrected bug in `cpp_stub.cpp` in the map of `ddsdde` to `DDSDDE` due to using `spatialDimensions` instead
  of `NTENS` (:jira:`685`, :pull:`47`). By `Nathan Miller`_.

Documentation
=============
- Add camelCase project name replacement instructions to project setup. By `Kyle Brindley`_.


******************
0.1.0 (2021-05-28)
******************

New Features
============
- Add CMake install configuration and CI/CD scripts for build, test, and installation to a Conda environment
  (:jira:`654`, :pull:`41`). By `Kyle Brindley`_.

Documentation
=============
- Update the Python package dependencies and add an example approach to future updates to the documentation
  (:jira:`636`, :pull:`37`). By `Kyle Brindley`_.
- Add file renaming commands to the project setup instructions (:jira:`634`, :pull:`38`). By `Kyle Brindley`_.
- Update the user manual to reflect required environment variable ``LD_LIBRARY_PATH`` (:jira:`662`, :pull:`43`). By
  `Kyle Brindley`_.

Internal Changes
================
- Update markdown syntax in README for wider compatibility (:jira:`604`, :pull:`36`). By `Kyle Brindley`_.
- Maintenance on ReST style guide updates (:jira:`604`, :pull:`36`). By `Kyle Brindley`_.
- Address BOOST output test stream deprecations and update minimum version
  (:jira:`654`, :pull:`41`). By `Kyle Brindley`_.
- Change project UMAT library name to avoid conflicts with external projects (:jira:`661`, :pull:`42`). By `Kyle
  Brindley`_.
- Remove the ``CXX`` compiler variable settings for build scripts (:jira:`671`, :pull:`44`). By `Kyle Brindley`_.

Enhancements
============
- Add multi-host and multi-environment CI/CD (:jira:`630`, :pull:`39`). By `Kyle Brindley`_.


******************
0.0.4 (2021-04-30)
******************

Documentation
=============
- Clarify behavior for custom target for the integration tests (:jira:`557`, :pull:`29`). By `Kyle Brindley`_.
- Add template documentation for the Abaqus material input definition (:jira:`575`, :pull:`31`). By `Kyle Brindley`_.
- Major overhaul of documentation organization to single source the Jenkins setup information from markdown files.  Adds
  the ``myst-parser`` Python package dependency and a pull request reviewer guide (:jira:`601`, :pull:`33`). By `Kyle
  Brindley`_.

Internal Changes
================
- Update Jenkins CI configuration to build and test for PRs to both ``master`` and ``dev`` branches (:jira:`544`,
  :pull:`26`). By `Kyle Brindley`_.
- Minor cleanup to root directory files. Move configuration and environment files to a subdirectory (:jira:`544`,
  :pull:`26`). By `Kyle Brindley`_.
- Add integration test CMake target for conditional rebuilds and file copy (:jira:`551`, :pull:`27`). By `Kyle
  Brindley`_.
- Add one ctest per abaqus input file (:jira:`551`, :pull:`27`). By `Kyle Brindley`_.
- Accept paths for input file in integration test shell script and check for errors in the abaqus stdout/stderr log
  (:jira:`551`, :pull:`27`). By `Kyle Brindley`_.
- Enable parallel CMake builds for continuous integration (CI) tests (:jira:`518`, :pull:`28`). By `Kyle Brindley`_.
- Add c++ source files ``*.cpp`` as dependencies for the Doxygen CMake target (:jira:`569`, :pull:`30`). By `Kyle
  Brindley`_.
- Add checks for ``STATEV`` and ``PROPS`` vector lengths to the abaqus interface. Throw exceptions with file and
  function name to interrupt Abaqus execution on input errors (:jira:`575`, :pull:`31`). By `Kyle Brindley`_.
- Add Abaqus interface unit tests for checking the ``STATEV`` and ``PROPS`` vector lengths (:jira:`575`, :pull:`31`). By
  `Kyle Brindley`_.
- Add unit tests for error codes in ``cpp_stub::sayHello`` (:jira:`334`, :pull:`32`). By `Kyle Brindley`_.

Enhancements
============
- Add error reporting to the Abaqus interface from the ``error_tools`` package (:jira:`334`, :pull:`32`). By `Kyle Brindley`_.


******************
0.0.3 (2021-04-13)
******************

Internal Changes
================
- Use ``abaqus_tools`` from a dedicated project (:jira:`535`, :pull:`23`). By `Kyle Brindley`_.
- Add ``bibtex_bibfiles`` variable to Sphinx configuration for newer version of ``sphinxcontrib.bibtex`` extension in
  Anaconda 2020 (:jira:`526`, :pull:`21`). By `Kyle Brindley`_.
- Add explicit list of documentation source files for better conditional CMake documentation re-builds (:jira:`526`,
  :pull:`21`). By `Kyle Brindley`_.


******************
0.0.2 (2021-02-11)
******************

Breaking changes
================

jenkins_deploy.sh
#!/usr/bin/env bash

set -x
# Override default permissions. Set group to rx with no write permissions.
umask 0022

date
script=`basename "$0"`
echo "Running ${script}"

# Parse branch basename from expected path-like name, e.g. origin/branch.
branch=$(basename ${GIT_BRANCH})

# Deploy master branch against release environment. All other branches against beta.
if [ ${branch} == master ]; then
    environment='release'
    env_alias='sv3r'
    master=true
elif [ ${branch} == dev ]; then
    environment='beta'
    env_alias='sv3b'
    master=false
else
    echo "Unexpected branch name. Exiting..."
    exit 1
fi

# Activate W-13 Python environment
case $(hostname) in
    sstelmo.lanl.gov|mayhem.lanl.gov)
        projects="/projects"
        module load python/2020.07-python-3.8
        ${env_alias}
        ;;  # No fall through
    sn-fey?.lanl.gov|sn-rfe?.lanl.gov|sn???.lanl.gov)
        projects="/usr/projects/ea"
        module load python/3.8-anaconda-2020.07
        source activate /usr/projects/ea/python/${environment}
        module load intel
        ;;
    *)
        echo "Unknown or unsupported host $(hostname)."
        exit 2
esac

# Can treat bash like a scripting language after conda activation
set -Eeuxo pipefail

# For master branch, update tags before building documentation and whl
if ${master}; then
    # Find cmake3 executable
    if [ -x "$(command -v cmake3)" ]; then
        cmake_exec=$(command -v cmake3)
    elif [ -x "$(command -v cmake)" ]; then
        cmake_exec=$(command -v cmake)
    else
        echo "Could not find cmake executable"
        exit 3
    fi
    # Build VERSION file from GetVersionFromGit.cmake without a full CMake configuration
    ${cmake_exec} -D PROJECT_NAME=cpp_stub -D VERSION_UPDATE_FROM_GIT=True -P src/cmake/GetVersionFromGitTag.cmake
    # GetVersionFromGit.cmake bumps micro/patch version. Retrieve next release from VERSION
    production_version=$(cut -f 1 -d '*' VERSION)
    developer_version=${production_version}+dev
    # Tag production commit and previous developer commit. Continue if already tagged.
    git tag -a ${production_version} -m "production release ${production_version}" || true
    last_merge_hash=$(git log --pretty=format:"%H" --merges -n 2 | tail -n 1)  # Assume last merge was dev->master. Pick previous
    git tag -a ${developer_version} -m "developer release ${developer_version}" ${last_merge_hash} || true
    git push origin --tags
fi

# Build project
./BUILD.sh

# Install project into conda environment by expected whl name.
./INSTALL.sh

# TODO: deploy documentation
